// nano-jsx/lib/ui/fab.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fab = void 0;
const component_1 = require("../component");
const core_1 = require("../core");
const _config_1 = require("./_config");
const _helpers_1 = require("./_helpers");
class Fab extends component_1.Component {
    render() {
        const { background = '#6200EE', color = 'white', extended = false, mini = false, center = false, left = false, onClick = () => { } } = this.props;
        const height = mini ? 40 : extended ? 48 : 56;
        const cssHash = core_1.strToHash(extended.toString() + mini.toString() + center.toString() + left.toString()); // Math.random().toString(36).substr(2, 9)
        const className = `fab-container-${cssHash}`;
        const styles = `
      .${className} {
        ${mini ? 'width: 40px;' : extended ? 'padding: 0px 12px;' : 'width: 56px;'}
        height: ${height}px;
        position: fixed;
        background: ${background};
        border-radius: 36px;
        display: flex;
        align-items: center;
        justify-content: center;
        color: ${color};
        cursor: pointer;

        z-index: ${_config_1.zIndex.fab}
        bottom: ${this.props.offsetY ? 16 + this.props.offsetY : 16}px;        
        ${left ? 'left: 16px;' : 'right: 16px;'}
        ${center ? 'transform: translateX(50%); right: 50%;' : ''}
        ${_config_1.boxShadow}
        ${_config_1.userSelect}
      }
    `;
        _helpers_1.addStylesToHead(styles, cssHash);
        const { children } = this.props;
        return core_1.h('div', { class: className, onClick: (e) => onClick(e) }, children);
    }
}
exports.Fab = Fab;
